### PR TITLE
Stop the crawler errors, beware the robot army

### PR DESCRIPTION
**Pull request by andycroll**

### CODE REVIEW

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+image:
+  base: '2018/stop-robots-crawlers-triggering-errors-rails'
+  alt: 'BB-8'
+  source: 'https://unsplash.com/photos/C8VWyZhcIIU'
+  credit: 'Joseph Chan'
+---
+
+When a Rails application can't find a record, it throws a 404 error. This is a standard HTTP code for browsers (and other clients) meaning 'not found'.
+
+When you have a Internet-facing site various search engines will be crawling your site. As you change things certain URLs might no longer work and the search engines can start generating a lot of errors by hitting all the links that used to exist.
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: *_an_ Internet-facing

Comment (nodunayo) [reply _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Generating a lot of errors where? In your monitoring setup?

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+
+## Instead of…
+
+Getting a bunch of unhelpful, distracting noise when Google (or another web crawler) hits deleted public pages…
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Again, where? It's getting to this point that made me go back and ask: wait a minute — where are we talking about?

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+           layout: nil,
+           status: :not_found
+  end
+end
+```
+
+
+## Use…
+
+A combination of the [`is_crawler`](https://github.com/ccashwell/is_crawler) gem and logic in your `application_controller.rb`.
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Choose one way: are you doing the "... <lowercase letter>" or "Capital"..when following on from the headings?

Do you know what I mean?

Comment (nodunayo) [reply _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Maybe: Use..the 'is_crawler' gem and required surrounding logic in your 'application_controller.rb'...

Line 44 is reading slightly confused...

Or maybe just: use the 'is_crawler' gem? After all, the example shows that you put the configuration in the application.rb?

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  def render_404
+    render file:   Rails.root.join('public', '404.html'),
+           layout: nil,
+           status: :not_found
+  end
+end
+```
+
+One of my sites gets hit by many more crawlers, that aren't included by default in the gem. So I add to the list of crawlers in an initializer.
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: *...by many crawlers that aren't checked for by default in the gem...

(Or maybe 'included' is still fine..)

Add a colon after 'initializer'.

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+Crawler::CUSTOM << Crawler.new(:dotbot, 'DotBot')
+Crawler::CUSTOM << Crawler.new(:mailru, 'Mail.RU_Bot')
+Crawler::CUSTOM << Crawler.new(:magestic12, 'MJ12bot')
+Crawler::CUSTOM << Crawler.new(:seznam, 'SeznamBot')
+```
+
+
+## But why?
+
+If you’re trying to make your app easier to maintain, it’s important to stay on top of your errors. You have probably wired your application up to an error reporting tool similar to [Rollbar](https://rollbar.com), [Honeybadger](https://honeybadger.io), [Bugsnag](https://bugsnag.com) or [Sentry](https://getsentry.com).
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Go on....comma after 'Bugsnag'. :P

Comment (nodunayo) [reply _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: *You have probably wired up your application to like Rollbar...

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+If you don’t know what are the real issues that your visitors are having and which are ‘noise’ from search engines you cannot focus and fix them. This approach means all the errors in your error provider should be _real_ and cleared out as much as possible - but that’s another article!
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: *If you cannot distinguish between the real issues that your visitors are having and the 'noise' from search engines, you cannot focus on the things that you should be fixing. You want all of the errors in your error provider/monitoring tool to be _real_..

What do you mean by 'cleared out' here? That you have the ability to work through them easily?

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+You likely don’t want to burn up all your credits with the tracking service, which you might if you’re receiving a large volume of errors.
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: *If you're receiving free credits from your tracking service, you'll burn through them if you're receiving a large volume of unnecessary errors.

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+It is possible to just ‘swallow’ all 404 errors, you most likely want the know when real users receive 404 pages, as it might indicate something important is broken.
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Hmm.

Maybe rephrase to get across: here's why it is important to know when real users receive 404 pages...

Though, I wonder if it is already clear or this should go elsewhere...I'll see when I read the next round.

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+
+## Why not?
+
+This might not be applicable to your domain. If most of your site is behind a sign in, this isn’t worth the extra effort.
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Maybe just take out the first sentence here and leave this as: If most of your site lives behind a sign in page, making this change isn't worth the extra effort.

**Review by nodunayo — DISMISSED** [review]

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+layout: article
+category: ruby
+image:
+  base: '2018/stop-robots-crawlers-triggering-errors-rails'
+  alt: 'BB-8'
+  source: 'https://unsplash.com/photos/C8VWyZhcIIU'
+  credit: 'Joseph Chan'
+---
+
+When a Rails application can't find a record, it throws a 404 error. This is a standard HTTP code for browsers (and other clients) meaning 'not found'.
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Get rid of the parentheses and the stuff in it...extra info that's not necessary I think...

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+When you have an Internet-facing site various search engines will be crawling your site. As you change things certain URLs might no longer work and the search engines can start generating a lot of ‘not found’ errors in your application by hitting all the links that used to exist.
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: *...various search engines will be crawling _it_. 

Then have a one-liner about what crawling is.

Comment (nodunayo) [reply _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Add a comma after "As you change things"

*As you change things, certain URLs might no longer work and the search engines/crawlers(?) can start generating a lot of 'not found' errors in your application when it hits them.

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+  def render_404
+    render file:   Rails.root.join('public', '404.html'),
+           layout: nil,
+           status: :not_found
+  end
+end
+```
+
+One of my sites gets hit by many more crawlers, that aren't included by default in the gem. So I add to the list of crawlers in an initializer:
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Take out the comma after 'crawlers'

*One of my sites...by many crawlers that aren't included by...

Comment (nodunayo) [reply _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: *So I add _these_ to the list of...

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+  Crawler::CUSTOM << Crawler.new(internal_name, agent_string)
+end
+```
+
+
+## But why?
+
+If you’re trying to make your app easier to maintain, it’s important to stay on top of your errors. You have probably wired up your application to an error monitoring tool similar to [Rollbar](https://rollbar.com), [Honeybadger](https://honeybadger.io), [Bugsnag](https://bugsnag.com), or [Sentry](https://getsentry.com).
+
+It is possible to just ‘swallow’ or ignore all 404 errors. However you want the know when real users receive 404 pages, as it might indicate something important is broken.
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: *It is tempting to ignore all 404 errors (?)

Comment (nodunayo) [reply _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: *..you want _to_ know

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+If you cannot distinguish between the genuine issues that your visitors are having and the ‘noise’ from search engines you cannot focus and fix _real_ problems.
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Comma after 'search engines'.

*...you cannot focus _on fixing_ real problems.

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+Eradicating these ‘robot errors’ should make fixing user-facing errors more straightforward. It's good to close and fix issues as they show up in your monitoring setup, but that’s another article!
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: I think take this paragraph out. You already explain why fixing real issues is easier without the noise above...then no point with the second sentence since you immediately say it's for another article.

**Review by nodunayo — DISMISSED** [review]

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md (hunk 15 above)
Comment (andycroll) [reply _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: Thought the same, rearranging the paragraph made this clearer.

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+image:
+  base: '2018/stop-robots-crawlers-triggering-errors-rails'
+  alt: 'BB-8'
+  source: 'https://unsplash.com/photos/C8VWyZhcIIU'
+  credit: 'Joseph Chan'
+---
+
+When a Rails application can't find a record, it throws a 404 error. This is a standard HTTP code for browsers meaning 'not found'.
+
+When you have an Internet-facing site various search engines will be crawling it. As you change things, certain URLs might change or cease to exist. This means search engines/crawlers can start generating a lot of ‘not found’ errors by trying to load pages that used to exist.
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: *in order for indexing purposes...

Not saying that's the wording — but isn't it worth saying what crawling means?

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+           layout: nil,
+           status: :not_found
+  end
+end
+```
+
+
+## Use…
+
+...the [`is_crawler`](https://github.com/ccashwell/is_crawler) gem and the configure it like so:
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: *_then_ configure...

File: _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+  def render_404
+    render file:   Rails.root.join('public', '404.html'),
+           layout: nil,
+           status: :not_found
+  end
+end
+```
+
+One of my sites gets hit by many more crawlers that aren't included by default in the gem. So I add these to the list of crawlers in an initializer:
```
Comment (nodunayo) [inline _posts/2018/2018-04-01-stop-robots-crawlers-triggering-errors-rails.md]: You either have to say something like: 'The 'is_crawler' gem comes with many default crawlers'..then the 'many more crawlers' phrasing works..

Or you change the sentence to something like: "One of my sites gets hit by many crawlers that aren't included.." Otherwise the 'more' comparison doesn't make sense.

**Review by nodunayo — DISMISSED** [review]
Almost there!! 😀

**Review by nodunayo — APPROVED** [review]
Looks good to me!
